presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushTopNThroughOuterJoin.java

```diff
@@ -0,0 +1,93 @@
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.TopNNode;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.List;
+
+import static com.facebook.presto.matching.Capture.newCapture;
+import static com.facebook.presto.spi.plan.TopNNode.Step.PARTIAL;
+import static com.facebook.presto.sql.planner.optimizations.QueryCardinalityUtil.isAtMost;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.LEFT;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+import static com.facebook.presto.sql.planner.plan.Patterns.Join.type;
+import static com.facebook.presto.sql.planner.plan.Patterns.TopN.step;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.planner.plan.Patterns.topN;
+
+/**
+ * Pushes a partial top N operation through a left or right outer join onto the preserved input.
+ * The ordering columns of the top N must be rewriteable to refer to only columns of
+ * the preserved input. Note that full outer joins are excluded as they
+ * can inject order affecting null rows into the join result.
+ */
+public class PushTopNThroughOuterJoin
+        implements Rule<TopNNode>
+{
+    private static final Capture<JoinNode> JOIN_CHILD = newCapture();
+
+    private static final Pattern<TopNNode> PATTERN =
+            topN().with(step().equalTo(PARTIAL))
+                    .with(source().matching(
+                            join().capturedAs(JOIN_CHILD).with(type().matching(type -> type == LEFT || type == RIGHT))));
+
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TopNNode parent, Captures captures, Context context)
+    {
+        JoinNode joinNode = captures.get(JOIN_CHILD);
+
+        List<VariableReferenceExpression> orderBySymbols = parent.getOrderingScheme().getOrderByVariables();
+
+        PlanNode left = joinNode.getLeft();
+        PlanNode right = joinNode.getRight();
+        JoinNode.Type type = joinNode.getType();
+
+        if ((type == LEFT)
+                && ImmutableSet.copyOf(left.getOutputVariables()).containsAll(orderBySymbols)
+                && !isAtMost(left, context.getLookup(), parent.getCount())) {
+            return Result.ofPlanNode(
+                    joinNode.replaceChildren(ImmutableList.of(
+                            parent.replaceChildren(ImmutableList.of(left)),
+                            right)));
+        }
+
+        if ((type == RIGHT)
+                && ImmutableSet.copyOf(right.getOutputVariables()).containsAll(orderBySymbols)
```

Comment:
Instead of repeating the code, just have two variables inner and outer, setting them appropriately based on join type so it's more robust.